File: app/src/main/java/xyz/rodit/snapmod/features/chatmenu/ToggleOption.java
package xyz.rodit.snapmod.features.chatmenu;

import xyz.rodit.snapmod.features.FeatureContext;
import xyz.rodit.snapmod.mappings.ActionMenuActionModel;
import xyz.rodit.snapmod.mappings.ActionMenuOptionTextViewModel;
import xyz.rodit.snapmod.mappings.ActionMenuOptionToggleItemViewModel;
import xyz.rodit.snapmod.mappings.SendChatAction;
import xyz.rodit.snapmod.mappings.SendChatActionDataModel;

public abstract class ToggleOption extends MenuPlugin {

    protected final int textResource;

    public ToggleOption(FeatureContext context, String name, int textResource) {
        super(context, name);
        this.textResource = textResource;
    }

    protected abstract boolean isToggled(String key);

    @Override
    protected Object createModel(String key) {
        SendChatActionDataModel actionDataModel = new SendChatActionDataModel(createEventData(name, key), false, null);
        SendChatAction action = new SendChatAction(actionDataModel);
        ActionMenuOptionTextViewModel textViewModel = new ActionMenuOptionTextViewModel(textResource, null, null, null, null, 62);
        return new ActionMenuOptionToggleItemViewModel(textViewModel, new ActionMenuActionModel(new Object[]{action.instance}), isToggled(key)).instance;
    }
}

File: app/src/main/java/xyz/rodit/snapmod/features/chatmenu/ToggleOption.kt
package xyz.rodit.snapmod.features.chatmenu

import xyz.rodit.snapmod.features.FeatureContext
import xyz.rodit.snapmod.mappings.*

abstract class ToggleOption(context: FeatureContext, name: String, private val textResource: Int) : MenuPlugin(context, name) {

    protected abstract fun isToggled(key: String?): Boolean

    override fun createModel(key: String?): Any {
        val actionDataModel = SendChatActionDataModel(createEventData(key!!), false, null)
        val action = SendChatAction(actionDataModel)
        val textViewModel = ActionMenuOptionTextViewModel(textResource, null, null, null, null, 62)
        return ActionMenuOptionToggleItemViewModel(
            textViewModel,
            ActionMenuActionModel(arrayOf(action.instance)),
            isToggled(key)
        ).instance
    }
}